### ISSUE TITLE
All are PVC disconnected from pods during upgrade

### ISSUE THREAD

**Issue by couloum** [kadalu/kadalu#606]
Hello,

We have performed the upgrade from `v0.8.3` to `v0.8.4` today in our preview environment. During the upgrade, all pods with a PVC attached got their PVC "disconnected". When we tried to access filesystem from the pod, we got this error: 

```
root@mysql-6f4d9fdcc5-8brk4:/# df -h
df: /var/lib/mysql: Transport endpoint is not connected
```

We had to delete all pods. After recreation, the PVC is correctly attached to new pods.

Is this the attended behavior?

Thank you

**Comment by couloum** [comment]
Hi,

As an additional note, I have observed that PVC get disconnected when the `csi-nodeplugin` instance running on the same node as the pod is being restarted.

Here are log of `csi-nodeplugin` on the same node as a pod that got it's transport endpoint disconnected:

```
# ktailf --tail 50 kadalu-csi-nodeplugin-mh8fg csi-node-driver-registrar
I0728 19:32:38.614787       1 main.go:112] Version: v2.0.1-0-g1aaa13a-dirty
I0728 19:32:38.614866       1 main.go:122] Attempting to open a gRPC connection with: "/plugin/csi.sock"
I0728 19:32:38.618185       1 connection.go:151] Connecting to unix:///plugin/csi.sock
I0728 19:32:39.637624       1 main.go:129] Calling CSI driver to discover driver name
I0728 19:32:39.637674       1 connection.go:180] GRPC call: /csi.v1.Identity/GetPluginInfo
I0728 19:32:39.637689       1 connection.go:181] GRPC request: {}
I0728 19:32:39.644735       1 connection.go:183] GRPC response: {"name":"kadalu","vendor_version":"0.8.4"}
I0728 19:32:39.644840       1 connection.go:184] GRPC error: <nil>
I0728 19:32:39.644854       1 main.go:139] CSI driver name: "kadalu"
I0728 19:32:39.644890       1 node_register.go:55] Starting Registration Server at: /registration/kadalu-reg.sock
I0728 19:32:39.645129       1 node_register.go:64] Registration Server started at: /registration/kadalu-reg.sock
I0728 19:32:39.645245       1 node_register.go:86] Skipping healthz server because port set to: 0
I0728 19:32:40.135968       1 main.go:79] Received GetInfo call: &InfoRequest{}
I0728 19:32:40.163772       1 main.go:89] Received NotifyRegistrationStatus call: &RegistrationStatus{PluginRegistered:true,Error:,}

# ktailf --tail 50 kadalu-csi-nodeplugin-mh8fg kadalu-nodeplugin
[2021-07-28 19:32:38,749] INFO [kadalulib - 339:start_all] - Started Process     name=csi
[2021-07-28 19:32:38,751] INFO [kadalulib - 339:start_all] - Started Process     name=metrics
[2021-07-28 19:32:38,753] INFO [kadalulib - 339:start_all] - Started Process     name=volumewatch
Starting watch on configmap
Setting up watches.
Watches established.
[2021-07-28 19:32:38,947] INFO [exporter - 85:<module>] - Started Kadalu Storage CSI Metrics exporter.   port=8000
[2021-07-28 19:32:39,011] INFO [main - 77:main] - Server started
[2021-07-28 19:36:43,186] INFO [nodeserver - 113:NodePublishVolume] - Mounted PV         volume=pvc-8aa6ec2f-3b83-4b2c-ae0d-78d4003773fa pvpath=subvol/d0/f9/pvc-8aa6ec2f-3b83-4b2c-ae0d-78d4003773fa pvtype=subvol hostvol=storage-pool duration_seconds=0.07492256164550781

[2021-07-28 19:36:43.171737 +0000] I [MSGID: 114057] [client-handshake.c:1126:select_server_supported_programs] 0-storage-pool-client-2: Using Program [{Program-name=GlusterFS 4.x v1}, {Num=1298437}, {Version=400}]
[2021-07-28 19:36:43.172460 +0000] I [MSGID: 114046] [client-handshake.c:855:client_setvolume_cbk] 0-storage-pool-client-1: Connected, attached to remote volume [{conn-name=storage-pool-client-1}, {remote_subvol=/bricks/storage-pool/data/brick}]
[2021-07-28 19:36:43.172488 +0000] I [MSGID: 108005] [afr-common.c:6046:__afr_handle_child_up_event] 0-storage-pool-replica-0: Subvolume 'storage-pool-client-1' came back up; going online.
[2021-07-28 19:36:43.172572 +0000] I [MSGID: 114046] [client-handshake.c:855:client_setvolume_cbk] 0-storage-pool-client-2: Connected, attached to remote volume [{conn-name=storage-pool-client-2}, {remote_subvol=/bricks/storage-pool/data/brick}]
[2021-07-28 19:36:43.172738 +0000] I [MSGID: 108002] [afr-common.c:6418:afr_notify] 0-storage-pool-replica-0: Client-quorum is met
[2021-07-28 19:36:43.172748 +0000] I [MSGID: 114046] [client-handshake.c:855:client_setvolume_cbk] 0-storage-pool-client-0: Connected, attached to remote volume [{conn-name=storage-pool-client-0}, {remote_subvol=/bricks/storage-pool/data/brick}]
[2021-07-28 19:36:43.175294 +0000] I [fuse-bridge.c:5318:fuse_init] 0-glusterfs-fuse: FUSE inited with protocol versions: glusterfs 7.24 kernel 7.31
[2021-07-28 19:36:43.175338 +0000] I [fuse-bridge.c:5950:fuse_graph_sync] 0-fuse: switched to graph 0
[2021-07-28 19:36:44.097418 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-07-28 19:36:44.104907 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-07-28 19:36:52.115477 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-07-28 19:37:12.921823 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-07-28 19:37:22.102407 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-07-28 19:37:22.121442 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-07-28 19:37:32.120099 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
[2021-07-28 19:37:32.170559 +0000] E [fuse-helpers.c:201:frame_fill_groups] 0-fuse: failed to open /proc/0/status: No such file or directory
.....
```

The error with `0-fuse: failed to open /proc/0/status` appears continuously in loop every second.

**Comment by amarts** [comment]
Thanks for these hints.

Thinking about it, makes sense to not restart nodeplugin unless people are ready to stop/drain all the pods from that node. Considering PVs are bind mounted into the app pods, there may be an issue where taking source mount down causes issues.

During upgrade, only the 'server' components should get upgraded, and that shouldn't disturb anything on the app/client side.

@aravindavk any concerns you see here?

**Comment by aravindavk** [comment]
> Thanks for these hints.
> 
> Thinking about it, makes sense to not restart nodeplugin unless people are ready to stop/drain all the pods from that node. Considering PVs are bind mounted into the app pods, there may be an issue where taking source mount down causes issues.
> 
> During upgrade, only the 'server' components should get upgraded, and that shouldn't disturb anything on the app/client side.
> 
> @aravindavk any concerns you see here?

If anything changed in Gluster client packages or node plugin code.  We may have to restart the node plugin during upgrade.

Good thing from the above log is `NodePublishVolume` is called after restart. The behaviour is same as mounting the PV for first time. Are you suspecting any issues with bind mount? or issue with Gluster client that is not able to connect back to the servers.

**Comment by amarts** [comment]
I am suspecting the issue with bind mounts... which is causing the `/proc/0/status` issue. Ie, pid is coming as 0, where as it shouldn't.

Need to understand this bit better to resolve this.

**Comment by amarts** [comment]
This got addressed as part of 0.8.6 release, where the node-plugin upgrade wont happen automatically. Thus no issues of PVC disconnect.
